planet/api/models.py

```diff
@@ -180,105 +134,277 @@ def name(self):
-        :type overwrite: bool, optional
-        :param callback: A function handle of the form
-            ``callback(start, wrote, total, finish, skip)`` that receives write
-            progress. Defaults to None
-        :type callback: function, optional
+        class _LOG():
+            def __init__(self, total, unit, filename, disable):
+                self.total = total
+                self.unit = unit
+                self.disable = disable
+                self.previous = 0
+                self.filename = filename
+
+                if not self.disable:
+                    LOGGER.debug(f'writing to {self.filename}')
+
+            def update(self, new):
+                if new-self.previous > self.unit and not self.disable:
+                    # LOGGER.debug(f'{new-self.previous}')
+                    perc = int(100 * new / self.total)
+                    LOGGER.debug(f'{self.filename}: '
+                                 f'wrote {perc}% of {self.total}')
+                    self.previous = new
+
+        unit = 1024*1024
+
+        mode = 'wb' if overwrite else 'xb'
+        try:
+            with open(filename, mode) as fp:
+                _log = _LOG(self.size, 16*unit, filename, disable=progress_bar)
+                with tqdm(total=self.size, unit_scale=True,
+                          unit_divisor=unit, unit='B',
+                          desc=filename, disable=not progress_bar) as progress:
+                    previous = self.num_bytes_downloaded
+                    async for chunk in self.aiter_bytes():
+                        fp.write(chunk)
+                        new = self.num_bytes_downloaded
+                        _log.update(new)
+                        progress.update(new-previous)
+                        previous = new
+        except FileExistsError:
+            LOGGER.info(f'File {filename} exists, not overwriting')
+
+
+def _get_filename_from_headers(headers):
+    """Get a filename from the Content-Disposition header, if available.
+
+    :param headers dict: a ``dict`` of response headers
+    :returns: a filename (i.e. ``basename``)
+    :rtype: str or None
+    """
+    cd = headers.get('content-disposition', '')
+    match = re.search('filename="?([^"]+)"?', cd)
+    return match.group(1) if match else None
+
+
+def _get_filename_from_url(url):
+    """Get a filename from a URL.
+
+    :returns: a filename (i.e. ``basename``)
+    :rtype: str or None
+    """
+    path = url.path
+    name = path[path.rfind('/')+1:]
+    return name or None
+
+
+def _get_random_filename(content_type=None):
+    """Get a pseudo-random, Planet-looking filename.
+
+    :returns: a filename (i.e. ``basename``)
+    :rtype: str
+    """
+    extension = mimetypes.guess_extension(content_type or '') or ''
+    characters = string.ascii_letters + '0123456789'
+    letters = ''.join(random.sample(characters, 8))
+    name = 'planet-{}{}'.format(letters, extension)
+    return name
+
+
+class Paged():
```

Comment:
Are there some concrete examples available for usage of this class? I'm trying to visualize a user workflow leveraging `Paged`, `Order`, and `Orders` from this module. :)

Reply:
A user will work with `Order`. All clients will receive paged responses. They will mostly be the same except they keys for e.g. next link will vary due to variations in the paged responses by the Data, Orders, etc APIs. So `Paged` captures the joint functionality, `Orders` specifies the special Orders API - specific functionality. These two classes will make more sense when the Data API client is implemented. `Orders` is used by `OrdersClient`. I don't expect `Orders` or `Paged` to be a part of a user workflow.